src/main/kotlin/org/dashevo/dpp/statetransition/StateTransition.kt

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2018-present, Dash Core Team
+ *
+ * This source code is licensed under the MIT license found in the
+ * COPYING file in the root directory of this source tree.
+ */
+
+package org.dashevo.dpp.statetransition
+
+import org.bitcoinj.core.*
+import org.bitcoinj.params.EvoNetParams
+import org.dashevo.dpp.BaseObject
+import org.dashevo.dpp.identity.IdentityPublicKey
+import org.dashevo.dpp.statetransition.errors.InvalidSignaturePublicKeyError
+import org.dashevo.dpp.statetransition.errors.InvalidSignatureTypeError
+import org.dashevo.dpp.statetransition.errors.PublicKeyMismatchError
+import org.dashevo.dpp.statetransition.errors.StateTransitionIsNotSignedError
+import org.dashevo.dpp.toBase64
+import org.dashevo.dpp.util.HashUtils
+import java.lang.Exception
+
+abstract class StateTransition(var signaturePublicKeyId: Int?,
+                               var signature: String?, var type: Types, var protocolVersion: Int = 0) : BaseObject() {
+
+    enum class Types(val type: Int) {
+        DATA_CONTRACT(1),
+        DOCUMENTS(2),
+        IDENTITY_CREATE(3),
+        IDENTITY_TOPUP(4),
+        IDENTITY_UPDATEKEY(5),
+        IDENTITY_CLOSEACCOUNT(6);
+    }
+
+    constructor(type: Types, protocolVersion: Int = 0) : this(null, null, type, protocolVersion)
+
+    override fun toJSON(): Map<String, Any> {
+        return toJSON(false)
+    }
+
+    open fun toJSON(skipSignature: Boolean): Map<String, Any> {
+        val json = hashMapOf<String, Any>()
+        json["protocolVersion"] = protocolVersion
+        json["type"] = type
+        if(!skipSignature) {
+            json["signature"] = if (signature != null) signature!! else ""
+            json["signaturePublicKeyId"] = if (signaturePublicKeyId != null) signaturePublicKeyId!! else ""
+        }
+        return json
+    }
+
+    fun serialize(skipSignature: Boolean): ByteArray {
+        return HashUtils.encode(this.toJSON(skipSignature))
+    }
+
+    fun sign(identityPublicKey: IdentityPublicKey, privateKey: String) {
+        val data = serialize(true)
```

Comment:
We don't have any tests to see if serialize() matches js-dpp or dashSync.  If this doesn't work our signatures won't match the other platforms.

Reply:
There is now a test successfully checks serialize().  We still don't have a signature check test